### PR TITLE
meta: Validate common ID against appstream metadata

### PR DESCRIPTION
**Pull request by cmatsuoka**
Appstream common IDs should reference IDs actually used in the metadata.
To prevent typos, add a consistency check and fail if any common ID is
not found in appstream files.

LP: #1814902

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `./runtests.sh static`?
- [x] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/internal/meta/_snap_packaging.py [hunk 1]
```diff
@@ -737,6 +738,28 @@ def _apply_passthrough(
...
+        common_id_list = self._extracted_metadata.common_id_list
+        for app in self._config_data["apps"]:
+            app_common_id = self._config_data["apps"][app].get("common-id")
+            if app_common_id not in common_id_list:
+                app_name = _get_app_name_from_common_id(
+                    self._config_data, app_common_id
+                )
+                logger.warning(
+                    "Common ID {common_id!r} specified in app {app!r} is "
+                    "not used in any appstream metafile.".format(
```
Comment (sergiusens) [inline snapcraft/internal/meta/_snap_packaging.py]: this is very specific to appstream, can you generalize the message?

File: tests/unit/test_meta.py [hunk 2]
```diff
@@ -1963,6 +1963,88 @@ def test_proper_command_chain(self):
...
+            part.build()
+            part.stage()
+            part.prime()
+        _snap_packaging.create_snap_packaging(config)
+        self.assertNotIn(
+            "is not used in any appstream metafile.", self.fake_logger.output
+        )
+
+    def test_common_id_mismatch(self):
+        config = self.make_snapcraft_project(common_id="tset.id.2")
```
Comment (sergiusens) [inline tests/unit/test_meta.py]: took me a second (or more I admit) to spot the typo, mind making common_id here something obviously wrong? tests.id.mismatch comes to mind

File: snapcraft/internal/meta/_snap_packaging.py [hunk 3]
```diff
@@ -737,6 +738,28 @@ def _apply_passthrough(
...
+            or not self._extracted_metadata.common_id_list
+            or "apps" not in self._config_data
+        ):
+            return
+
+        common_id_list = self._extracted_metadata.common_id_list
+        for app in self._config_data["apps"]:
+            app_common_id = self._config_data["apps"][app].get("common-id")
+            if app_common_id not in common_id_list:
+                app_name = _get_app_name_from_common_id(
```
Comment (sergiusens) [inline snapcraft/internal/meta/_snap_packaging.py]: isn't app_name just app in this list?

File: snapcraft/internal/meta/_snap_packaging.py (hunk 1 above)
Comment (cmatsuoka) [reply snapcraft/internal/meta/_snap_packaging.py]: Message changed.

File: tests/unit/test_meta.py (hunk 2 above)
Comment (cmatsuoka) [reply tests/unit/test_meta.py]: The subtle typo was a terrible choice for the test. Changing to something obviously different.

File: snapcraft/internal/meta/_snap_packaging.py (hunk 3 above)
Comment (cmatsuoka) [reply snapcraft/internal/meta/_snap_packaging.py]: Yes, it is. The convoluted data flow was a leftover from a refactoring, fixing that.

**Comment by sergiusens** [comment]
you probably want to strip the logger output

**Comment by cmatsuoka** [comment]
> you probably want to strip the logger output

Could you elaborate a bit more on that? I think I lost the context here :)

File: tests/unit/test_meta.py [hunk 4]
```diff
@@ -1963,6 +1963,88 @@ def test_proper_command_chain(self):
...
+        for part in config.parts.all_parts:
+            part.makedirs()
+            part.mark_pull_done()
+            part.build()
+            part.stage()
+            part.prime()
+        _snap_packaging.create_snap_packaging(config)
+        self.assertIn(
+            "Common ID 'test.id.mismatch' specified in app 'test-app' is not used in any appstream metafile.",
+            self.fake_logger.output,
```
Comment (sergiusens) [inline tests/unit/test_meta.py]:
```suggestion
            self.fake_logger.output.strip(),
```

Comment (sergiusens) [reply tests/unit/test_meta.py]: this is why your test fails

Comment (cmatsuoka) [reply tests/unit/test_meta.py]: Agh, sorry, I didn't see that. I wonder why the unit tests ran successfully locally.

**Pull request merged**